File: concept/15_2D_array/spiral_print/spiral_print.cpp
#include <iostream>
#include <vector>
using namespace std;

int main()
{
    int arr[3][3] = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};

    for (int i = 0; i < 3; i++)
    {
        for (int j = 0; j < 3; j++)
        {
            cout << arr[i][j] << " ";
        }
        cout << endl;
    }

    int row = 3;
    int column = 3;
    int count = 0;
    int total = row * column;

    int startRow = 0;
    int startColumn = 0;
    int endRow = row - 1;
    int endColumn = column - 1;

    vector<int> v;

    while (count < total)
    {
        // startRow
        for (int i = startColumn; count < total && i <= endColumn; i++)
        {
            v.push_back(arr[startRow][i]);
            count++;
        }
        startRow++;

        // endColumn
        for (int i = startRow; count < total && i <= endRow; i++)
        {
            v.push_back(arr[i][endColumn]);
            count++;
        }
        endColumn--;

        // End Row
        for (int i = endColumn; count < total && i >= startColumn; i--)
        {
            /* code */
            v.push_back(arr[endRow][i]);
            count++;
        }
        endRow--;

        // startColumn
        for (int i = endRow; count < total && i >= startRow; i--)
        {
            /* code */
            v.push_back(arr[i][startColumn]);
            count++;
        }
        startColumn++;
    }
    for (int i = 0; i < v.size(); i++)
    {
        cout << v[i] << " ";
    }
}

File: concept/15_2D_array/spiral_print/spiral_print.dart
void main() {
  List<List<int>> data = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
  ];

  for (var v in data) {
    print(v);
  }
  int r = 3;
  int c = 3;
  int count = 0;
  int total = r * c;

  int starRow = 0;
  int starCol = 0;
  int endRow = r - 1;
  int endCol = c - 1;

  List<int> newData = [];

  while (count < total) {
    // Start Row
    for (var i = starCol; count < total && i <= endCol; i++) {
      int value = data[starRow][i];
      newData.add(value);
      count++;
    }
    starRow++;

    for (var i = starRow; count < total && i <= endRow; i++) {
      int value = data[i][endCol];
      newData.add(value);
      count++;
    }
    endCol--;

    for (var i = endCol; count < total && i >= starCol; i--) {
      int value = data[endRow][i];
      newData.add(value);
      count++;
    }
    endRow--;

    for (var i = endRow; count < total && i >= starRow; i--) {
      int value = data[i][starCol];
      newData.add(value);
      count++;
    }
    starCol++;
  }

  print(newData);
}
